component/xload/xload.go

```
@@ -0,0 +1,300 @@
+	if err != nil {
+		log.Err("Xload::Configure : Failed to parse mode %s [%s]", conf.Mode, err.Error())
+		return fmt.Errorf("invalid mode in xload : %s", conf.Mode)
+	}
+
+	if mode == EMode.INVALID_MODE() {
+		log.Err("Xload::Configure : Invalid mode : %s", conf.Mode)
+		return fmt.Errorf("invalid mode in xload : %s", conf.Mode)
+	}
+
+	xl.mode = mode
+
+	return nil
+}
+
+// Start : Pipeline calls this method to start the component functionality
+func (xl *Xload) Start(ctx context.Context) error {
+	log.Trace("Xload::Start : Starting component %s", xl.Name())
+
+	xl.workerCount = MAX_WORKER_COUNT
+	xl.blockPool = NewBlockPool(xl.blockSize, xl.workerCount*3)
+	if xl.blockPool == nil {
+		log.Err("Xload::Start : Failed to create block pool")
+		return fmt.Errorf("failed to create block pool")
+	}
+
+	var err error
+
+	// Xload : start code goes here
+	switch xl.mode {
+	case EMode.CHECKPOINT():
+		// Start checkpoint thread here
+		return fmt.Errorf("checkpoint is currently unsupported")
+	case EMode.DOWNLOAD():
+		// Start downloader here
+		err = xl.startDownloader()
+		if err != nil {
+			log.Err("Xload::Start : Failed to start downloader [%s]", err.Error())
+			return err
+		}
+	case EMode.UPLOAD():
+		// Start uploader here
+		return fmt.Errorf("uploader is currently unsupported")
+	case EMode.SYNC():
+		//Start syncer here
+		return fmt.Errorf("sync is currently unsupported")
+	default:
+		log.Err("Xload::Start : Invalid mode : %s", xl.mode.String())
+		return fmt.Errorf("invalid mode in xload : %s", xl.mode.String())
+	}
+
+	return xl.startComponents()
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (xl *Xload) Stop() error {
+	log.Trace("Xload::Stop : Stopping component %s", xl.Name())
+
+	xl.comps[0].stop()
+	xl.blockPool.Terminate()
+
+	// TODO:: xload : should we delete the files from local path
+	err := common.TempCacheCleanup(xl.path)
+	if err != nil {
+		log.Err("unable to clean xload local path [%s]", err.Error())
```

Should we return an error from this function? If not, can we remove the return type entirely since the function currently only returns `nil`.

---

Returning error here